lib/kube/proxy/transport.go

```diff
@@ -0,0 +1,489 @@
+		return sess.kubeAPICreds.getTransport(sess.upgradeToHTTP2), nil
+	} else if f.cfg.ReverseTunnelSrv != nil {
+		// If agent is running in proxy mode, create a new transport for the local cluster.
+		httpTransport, err = f.newLocalClusterTransport(sess.kubeClusterName)
+	} else {
+		return nil, trace.BadParameter("no reverse tunnel server or credentials provided")
+	}
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Cache the transport.
+	f.cachedTransport.Set(key, httpTransport, 5*time.Hour)
+	// Return the transport depending on whether HTTP/2 is enabled.
+	// Distinction is made because the SPDY protocol is not supported by HTTP/2
+	// and we must use HTTP/1.1 for it.
+	if sess.upgradeToHTTP2 {
+		return httpTransport.h2Transport, nil
+	}
+	return httpTransport.h1Transport, nil
+}
+
+// transportCacheKey returns a key used to cache transports.
+// If the cluster is remote, the key is the teleport cluster name.
+// If the cluster is local, the key is the teleport cluster name and the kubernetes
+// cluster name.
+func transportCacheKey(sess *clusterSession) string {
+	if sess.teleportCluster.isRemote {
+		return sess.teleportCluster.name
+	}
+	return filepath.Join(sess.teleportCluster.name, sess.kubeClusterName)
+}
+
+// wrapTransport wraps the provided transport with the Kubernetes transport config
+// if it is not nil.
+func wrapTransport(rt http.RoundTripper, transportConfig *transport.Config) (http.RoundTripper, error) {
+	if transportConfig == nil {
+		return rt, nil
+	}
+
+	wrapped, err := transport.HTTPWrappersForConfig(transportConfig, rt)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return enforceCloseIdleConnections(wrapped, rt), nil
+}
+
+// newTransport creates a new [http.Transport] with the provided dialer and TLS
+// config.
+// The transport is configured to use a connection pool and to close idle
+// connections after a timeout.
+func newTransport(dial dialContextFunc, tlsConfig *tls.Config) *http.Transport {
+	return &http.Transport{
+		DialContext:     dial,
+		TLSClientConfig: tlsConfig,
+		// Increase the size of the connection pool. This substantially improves the
+		// performance of Teleport under load as it reduces the number of TLS
+		// handshakes performed.
+		MaxIdleConns:        defaults.HTTPMaxIdleConns,
+		MaxIdleConnsPerHost: defaults.HTTPMaxIdleConnsPerHost,
+		// IdleConnTimeout defines the maximum amount of time before idle connections
+		// are closed. Leaving this unset will lead to connections open forever and
+		// will cause memory leaks in a long running process.
+		IdleConnTimeout: defaults.HTTPIdleTimeout,
+	}
+}
+
+// versionWithoutImpersonation is the version of Teleport that starts supporting
+// impersonation. Before this version, the client will not use impersonation.
+var versionWithoutImpersonation = semver.New(utils.VersionBeforeAlpha("13.0.0"))
+
+// teleportVersionInterface is an interface that allows to get the Teleport version of
+// a server.
+// DELETE IN 14.0.0
```

Comment:
It won't be a problem.

the root proxy will dial to the leaf proxy using the new impersonated method. Leaf proxy will dial to the kubernetes service using the old method since kubernetes service does not support it.

Reply:
The issue is that I don't think that we specify how old the components of the leaf cluster can be; if we interpret "leaf cluster version" to mean both auth and proxy (which is the sensible interpretation, I think) then you're right, we can delete it in v14.

The alternative would be that the leaf auth is one version behind and then the leaf proxy is one version behind its auth, but that would mean that the remote cluster reverse tunnel needs to stay compatible for two major versions of the proxy, which is kind of a tall order.